cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -0,0 +1,268 @@
+   * @param timeRangeStart Start of the time range that the scan should begin from
+   * @param timeRangeEnd End of the time range that the scan should end at
+   * @return List of WorkflowRunRecords
+   * @throws Exception
+   */
+  public List<WorkflowRunRecord> scan(Id.Workflow id, long timeRangeStart, long timeRangeEnd) throws Exception {
+    byte[] startRowKey = new MDSKey.Builder().add(id.getApplication().getNamespaceId()).add(id.getApplicationId()).
+      add(id.getId()).add(timeRangeStart).build().getKey();
+    byte[] endRowKey = new MDSKey.Builder().add(id.getApplication().getNamespaceId()).add(id.getApplicationId()).
+      add(id.getId()).add(timeRangeEnd).build().getKey();
+    Scan scan = new Scan(startRowKey, endRowKey);
+
+    Scanner scanner = table.scan(scan);
+    Row indexRow;
+    List<WorkflowRunRecord> workflowRunRecordList = new ArrayList<>();
+    while ((indexRow = scanner.next()) != null) {
+      Map<byte[], byte[]> columns = indexRow.getColumns();
+      String workflowRunId = Bytes.toString(columns.get(RUNID));
+      long timeTaken = Bytes.toLong(columns.get(TIME_TAKEN));
+
+      List<ActionRun> actionRunList = GSON.fromJson(Bytes.toString(columns.get(NODES)), ACTION_RUNS_TYPE);
+      WorkflowRunRecord workflowRunRecord = new WorkflowRunRecord(workflowRunId, timeTaken, actionRunList);
+      workflowRunRecordList.add(workflowRunRecord);
+    }
+    return workflowRunRecordList;
+  }
+
+  /**
+   * This method returns the statistics for a corresponding workflow. The user has to
+   * provide a time interval and a list of percentiles that are required.
+   *
+   * @param id The workflow id
+   * @param startTime The start of the time range from where the user wants the statistics
+   * @param endTime The end of the time range till where the user wants the statistics
+   * @param percentiles The list of percentiles that the user wants information on
+   * @return A statistics object that provides information about the workflow or null if there are no runs associated
+   * with the workflow
+   * @throws Exception
+   */
+  @Nullable
+  public WorkflowStatistics getStatistics(Id.Workflow id, long startTime,
+                                          long endTime, List<Double> percentiles) throws Exception {
+    List<WorkflowRunRecord> workflowRunRecords = scan(id, startTime, endTime);
+    int count = workflowRunRecords.size();
+
+    if (count == 0) {
+      return null;
+    }
+
+    double avgRunTime = 0.0;
+    for (WorkflowDataset.WorkflowRunRecord workflowRunRecord : workflowRunRecords) {
+      avgRunTime += workflowRunRecord.getTimeTaken();
+    }
+    avgRunTime /= count;
+
+    workflowRunRecords = sort(workflowRunRecords);
+
+    List<PercentileInformation> percentileInformationList =
+      getPercentiles(workflowRunRecords, percentiles);
+
+    Map<String, List<Long>> actionToRunRecord = getActionRuns(workflowRunRecords);
+
+    Map<String, Map<String, Double>> actionToStatistic = new HashMap<>();
```

Comment:
@gsps1 suggested to use a POJO here since the keys in the map are not dynamic and few in count, right?